### PR TITLE
fix(hmr-plugin): improved hmr update state

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Update before ngxsOnInit

## What is the new behavior?

Update after ngxsAfterBootstrap

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/hmr-plugin/src/internal/common.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+export function getStateFromHmrStorage<S = NgxsStoreSnapshot>(): Partial<S> {
+  return JSON.parse(sessionStorage.getItem(NGXS_HMR_SNAPSHOT_KEY) || '{}');
+}
+
+export function setStateInHmrStorage<S = NgxsStoreSnapshot>(state: S): void {
+  return sessionStorage.setItem(NGXS_HMR_SNAPSHOT_KEY, JSON.stringify(state));
+}
+
+export function getStateStream<T>(ref: NgModuleRef<T>): StateStream | null {
+  return ref.injector.get(Store, null)['_stateStream'] || null;
```
Comment (arturovt) [inline packages/hmr-plugin/src/internal/common.ts]: If injector's get returns `null` - it will throw `Cannot read property '_stateStream' of null`

```TS
const store = ref.injector.get<Store>(Store, null);
if (store) {
  return store['_stateStream'];
}

return null;
```

File: packages/hmr-plugin/src/internal/common.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+export function getBootstrap<T>(ref: NgModuleRef<T>): NgxsBootstrapper | null {
+  return ref.injector.get(NgxsBootstrapper, null) || null;
```
Comment (arturovt) [inline packages/hmr-plugin/src/internal/common.ts]: `return ref.injector.get(NgxsBootstrapper, null)`
no need for `|| null`

File: packages/hmr-plugin/src/internal/state-event-loop.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import { getStateContext } from './state-context';
+import { CallStackFrame, NgxsHmrLifeCycle } from '../symbols';
+import { getBootstrap, getStateFromHmrStorage, getStateStream } from './common';
+
+export function stateEventLoop<T extends NgxsHmrLifeCycle<S>, S>(
+  ref: NgModuleRef<T>,
+  frame: CallStackFrame<S>
+) {
+  const stateContext: StateContext<S> | undefined = getStateContext<T, S>(ref);
+  if (stateContext) {
```
Comment (arturovt) [inline packages/hmr-plugin/src/internal/state-event-loop.ts]:
```TS
if (!stateContext) {
  return;
}
```

No need for nesting

File: packages/hmr-plugin/src/internal/state-event-loop.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    const previousState: Partial<S> = getStateFromHmrStorage<S>();
+    const existSavedState: boolean = Object.keys(previousState).length > 0;
+
+    // get event subscriptions
+    const stateStream: StateStream | null = getStateStream<T>(ref);
+    const bootstrap: NgxsBootstrapper | null = getBootstrap(ref);
+    const canBeUpdateState: boolean = Boolean(existSavedState && stateStream && bootstrap);
+
+    if (canBeUpdateState) {
```
Comment (arturovt) [inline packages/hmr-plugin/src/internal/state-event-loop.ts]:
```TS
if (!canBeUpdateState) {
  return;
}
```

File: packages/hmr-plugin/src/internal/state-context.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import { isStateOperator } from '@ngxs/store/operators';
+
+import { NgxsHmrLifeCycle, NgxsStoreSnapshot } from '../symbols';
+
+export function getStateContext<T extends NgxsHmrLifeCycle<S>, S = NgxsStoreSnapshot>(
+  ref: NgModuleRef<T>
+): StateContext<S> | undefined {
+  const store: Store = ref.injector.get(Store, null);
+
+  if (store) {
```
Comment (arturovt) [inline packages/hmr-plugin/src/internal/state-context.ts]:
```TS
if (!store) {
  return;
}
```

**Comment by markwhitfeld** [comment]
I'm still trying to catch up. Will look at this tomorrow.

File: packages/hmr-plugin/src/internal/state-event-loop.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import { getBootstrap, getStateFromHmrStorage, getStateStream } from './common';
+
+export function stateEventLoop<T extends NgxsHmrLifeCycle<S>, S>(
+  ref: NgModuleRef<T>,
+  frame: CallStackFrame<S>
+) {
+  const stateContext: StateContext<S> | undefined = getStateContext<T, S>(ref);
+  if (stateContext) {
+    const previousState: Partial<S> = getStateFromHmrStorage<S>();
+    const existSavedState: boolean = Object.keys(previousState).length > 0;
```
Comment (markwhitfeld) [inline packages/hmr-plugin/src/internal/state-event-loop.ts]:
```suggestion
    const savedStateExists: boolean = Object.keys(previousState).length > 0;
```
Proposed variable rename.

File: packages/hmr-plugin/src/internal/state-event-loop.ts (hunk 4 above)
Comment (markwhitfeld) [reply packages/hmr-plugin/src/internal/state-event-loop.ts]: Also I think a better variable name would be `canUpdateState`.

File: packages/hmr-plugin/src/internal/state-event-loop.ts (hunk 6 above)
Comment (arturovt) [reply packages/hmr-plugin/src/internal/state-event-loop.ts]: @markwhitfeld maybe just `const savedStateExists: boolean = Object.keys(previousState).length`, I mean without `> 0`?

Comment (markwhitfeld) [reply packages/hmr-plugin/src/internal/state-event-loop.ts]: Then its type would not be boolean. I think it is best to leave the `> 0` part there.

File: packages/hmr-plugin/src/internal/common.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { NgModuleRef } from '@angular/core';
+import { NgxsBootstrapper, StateStream, Store } from '@ngxs/store';
```
Comment (markwhitfeld) [inline packages/hmr-plugin/src/internal/common.ts]: NgxsBootStrapper has now moved to the `internal` subpackage

File: integration/app/app-routing.module.ts [hunk 8]
```diff
@@ -6,7 +6,8 @@ import { NgModule } from '@angular/core';
     RouterModule.forRoot([
       { path: '', pathMatch: 'full', redirectTo: '/list' },
       { path: 'list', loadChildren: '@integration/list/list.module#ListModule' },
-      { path: 'detail', loadChildren: '@integration/detail/detail.module#DetailModule' }
+      { path: 'detail', loadChildren: '@integration/detail/detail.module#DetailModule' },
+      { path: 'counter', loadChildren: '@integration/counter/counter.module#CounterModule' }
```
[the remark this answers is not in the document]
Comment (eranshmil) [reply integration/app/app-routing.module.ts]: I agree.

File: integration/app/counter/counter.actions.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export class CounterStateChangeAction {
+  public static readonly type = 'CounterStateChangeAction';
```
Comment (eranshmil) [inline integration/app/counter/counter.actions.ts]: This should match the examples and the conventions in the docs.
```ts
export class ChangeAction {
  static readonly type = '[Counter] Change';
}
```

File: packages/hmr-plugin/src/hmr-bootstrap.ts [hunk 10]
```diff
@@ -1,7 +1,9 @@
 import { ApplicationRef, NgModuleRef } from '@angular/core';
-import { createNewHosts } from '@angularclass/hmr';
+import { createNewHosts, hmrModule } from '@angularclass/hmr';
 
-import { hmrInit, hmrDoBootstrap, hmrDoDispose } from './internal';
+import { hmrInit } from './internal/init';
```
Comment (eranshmil) [inline packages/hmr-plugin/src/hmr-bootstrap.ts]: Add `index.ts` file to preserve this imports.

File: packages/hmr-plugin/src/internal/on-destroy.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+import { getStateContext } from './state-context';
+import { NgxsHmrLifeCycle, NgxsStoreSnapshot } from '../symbols';
+
+export function hmrBeforeOnDestroy<T extends NgxsHmrLifeCycle<S>, S = NgxsStoreSnapshot>(
+  ref: NgModuleRef<T>
+): Partial<S> {
+  let resultSnapshot: Partial<S> = {};
+  const ngxsHmrLifeCycle = ref.instance;
+  const hmrNgxsStoreOnDestroyFn = ngxsHmrLifeCycle.hmrNgxsStoreBeforeOnDestroy;
```
Comment (eranshmil) [inline packages/hmr-plugin/src/internal/on-destroy.ts]: No need to create the variable `ngxsHmrLifeCycle`.

File: packages/hmr-plugin/src/internal/do-bootstrap.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+import { StateContext } from '@ngxs/store';
+
+import { setStateInHmrStorage } from './common';
+import { stateEventLoop } from './state-event-loop';
+import { HmrNgxsStoreOnInit, NgxsHmrLifeCycle, NgxsStoreSnapshot } from '../symbols';
+
+export function hmrDoBootstrap<T extends NgxsHmrLifeCycle<S>, S = NgxsStoreSnapshot>(
+  ref: NgModuleRef<T>
+): NgModuleRef<T> {
+  const ngxsHmrLifeCycle: NgxsHmrLifeCycle<S> = ref.instance;
```
Comment (eranshmil) [inline packages/hmr-plugin/src/internal/do-bootstrap.ts]: The same as the comment in `on-destroy.ts`.
Also, the typings should be removed/included in the other file, to keep consistency.

File: packages/hmr-plugin/src/internal/state-event-loop.ts (hunk 6 above)
Comment (eranshmil) [reply packages/hmr-plugin/src/internal/state-event-loop.ts]: > Then its type would not be boolean. I think it is best to leave the `> 0` part there.

So he can add `!!` in the beginning.

File: packages/hmr-plugin/src/internal/state-event-loop.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    if (canBeUpdateState) {
+      let eventId: number;
+      const waitTime = 100;
+
+      bootstrap!.appBootstrapped$.subscribe(() => {
+        const stateStreamSub: Subscription = stateStream!.subscribe(() => {
+          // setTimeout used for zone detection after set hmr state
+          clearInterval(eventId);
+          // only need to get the last message
+          eventId = window.setTimeout(() => {
```
Comment (eranshmil) [inline packages/hmr-plugin/src/internal/state-event-loop.ts]: Do we really need to use `setTimeout`?
Also, no need to write `window.`.

Comment (splincode) [reply packages/hmr-plugin/src/internal/state-event-loop.ts]: Yes, because we need triggers for NgZone

File: packages/hmr-plugin/src/internal/state-event-loop.ts (hunk 6 above)
Comment (markwhitfeld) [reply packages/hmr-plugin/src/internal/state-event-loop.ts]: You could do the `!!` but it offers no advantage over the `> 0` which actually describes what we are doing and requires no translation.

**Comment by splincode** [comment]
@markwhitfeld I added tests

File: integration/main.browser.ts [hunk 14]
```diff
@@ -13,7 +13,7 @@ const bootstrap = () => platformBrowserDynamic().bootstrapModule(AppBrowserModul
 
 if (environment.hmr) {
   if (module['hot']) {
-    hmrNgxsBootstrap(module, bootstrap).catch(err => console.log(err));
+    hmr({ module, bootstrap }).catch(err => console.error(err));
```
Comment (markwhitfeld) [inline integration/main.browser.ts]: Since module and bootstrap are not optional parameters and are critical for the hmr plugin to work these should not be part of the options but rather be the first 2 compulsory parameters.

**Comment by markwhitfeld** [comment]
I see that there are extensive changes so I will have to do a full review again. Hopefully tonight. In the mean time would it be possible to make the `module` and `bootstrap` hmr arguments parameters as opposed to part of the options object?

File: integration/main.browser.ts (hunk 14 above)
Comment (splincode) [reply integration/main.browser.ts]: done

**Comment by markwhitfeld** [comment]
Please wait for another approval before merging.
Just need to fix the failing build too though!

File: packages/hmr-plugin/src/symbols.ts [hunk 15]
```diff
@@ -1,22 +1,72 @@
...
+   * (default: 100ms)
    */
-  hmrNgxsStoreBeforeOnDestroy(context: StateContext<T>): T;
+  deferTime?: number;
+
+  /**
+   * @description - method to call intermediate result
+   * @param manager
+   */
+  hmrAfterOnInit?: HmrAfterOnInit<T, S>;
```
Comment (markwhitfeld) [inline packages/hmr-plugin/src/symbols.ts]: Why is this necessary to expose `HmrAfterOnInit` as part of the options?
This would cause the manager and much of the internals to become public.

Comment (splincode) [reply packages/hmr-plugin/src/symbols.ts]: `HmrAfterOnInit` - Because you want to control the process as a user.

That is, you want to understand when your method was executed to perform for example something else.

```ts
await hmr(module, bootstrap, { 
  hmrAfterOnInit: ({ store }) => {
    store.dispatch(new UPDATE_STATE_HMR()); // custom trigger all states if we want
  }
});
```

File: packages/hmr-plugin/src/internal/hmr-store-context.ts [hunk 16]
```diff
@@ -0,0 +1,46 @@
+import { StateContext, Store } from '@ngxs/store';
+import { isStateOperator } from '@ngxs/store/operators';
+import { NgModuleRef } from '@angular/core';
+import { NgxsBootstrapper } from '@ngxs/store/internals';
+
+export class HmrStoreContext<T, S> {
+  public store: Store;
+  public bootstrap: NgxsBootstrapper;
+
+  constructor(module: NgModuleRef<T>) {
```
[the remark this answers is not in the document]
Comment (splincode) [reply packages/hmr-plugin/src/internal/hmr-store-context.ts]: Unfortunately the magic of the anguly is not available here.

File: packages/hmr-plugin/src/internal/hmr-options-builder.ts [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+
+export class HmrOptionBuilder<T extends NgxsHmrLifeCycle<S>, S> {
+  public readonly deferTime: number;
+  public readonly autoClearLogs: boolean;
+  public readonly hmrAfterOnInit: HmrAfterOnInit<T, S>;
+
+  constructor({ deferTime, autoClearLogs, hmrAfterOnInit }: NgxsHmrOptions<T, S>) {
+    this.deferTime = deferTime || 100;
+    this.hmrAfterOnInit = hmrAfterOnInit || ((_: HmrManager<T, S>) => {});
+    this.autoClearLogs = autoClearLogs === undefined ? true : autoClearLogs;
```
[the remark this answers is not in the document]
Comment (splincode) [reply packages/hmr-plugin/src/internal/hmr-options-builder.ts]: Unfortunately, we do not need these logs, since they will be dumped from the old session

File: packages/hmr-plugin/src/internal/hmr-storage.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
+import { NGXS_HMR_SNAPSHOT_KEY } from '../symbols';
+
+export class HmrStorage<S> {
+  private readonly keyStore: string = NGXS_HMR_SNAPSHOT_KEY;
+
+  public validateExistHmrStorage(): void {
+    if (!this.existHmrStorage) {
```
Comment (arturovt) [inline packages/hmr-plugin/src/internal/hmr-storage.ts]: maybe its better to name variable `hmrStorageExists`?
and also
```TS
if (hmrStorageExists) {
  return;
}
```

File: packages/hmr-plugin/src/symbols.ts (hunk 15 above)
Comment (markwhitfeld) [reply packages/hmr-plugin/src/symbols.ts]: Could we then only expose a `stateContext` as a parameter to this method?
Not any internal HMR plugin objects.

Comment (splincode) [reply packages/hmr-plugin/src/symbols.ts]: then I don't know how to test it in my tests
https://github.com/ngxs/store/blob/fe4b138c1989ea667a3c1906f8d71c9c7624e1f2/packages/hmr-plugin/src/tests/hmr-manager.spec.ts

Comment (markwhitfeld) [reply packages/hmr-plugin/src/symbols.ts]: You shouldn't be testing internals but rather verifying the state of the system through public methods.

**Comment by markwhitfeld** [comment]
I just merged the reusable states PR. Was that the PR holding up your work on this one?

**Comment by splincode** [comment]
@markwhitfeld Yes, in fact, now you can finish this PR, and everything is completely ready for us))

**Comment by splincode** [comment]
Add action for hmr lifecycle
image

**Comment by splincode** [comment]
@markwhitfeld @arturovt @eranshmil At the moment I have completed work on the plugin

File: packages/hmr-plugin/src/symbols.ts (hunk 15 above)
Comment (splincode) [reply packages/hmr-plugin/src/symbols.ts]: done

File: packages/hmr-plugin/src/internal/hmr-store-context.ts [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+    this.store = store;
+    this.bootstrap = module.injector.get(NgxsBootstrapper);
+  }
+
+  /**
+   * @description
+   * must be taken out into  @ngxs/store/internals
+   */
+  public get stateContext(): StateContext<S> {
```
[the remark this answers is not in the document]
Comment (splincode) [reply packages/hmr-plugin/src/internal/hmr-store-context.ts]: done

File: packages/hmr-plugin/src/internal/hmr-store-context.ts [hunk 20]
```diff
@@ -0,0 +1,46 @@
+import { StateContext, Store } from '@ngxs/store';
+import { isStateOperator } from '@ngxs/store/operators';
+import { NgModuleRef } from '@angular/core';
+import { NgxsBootstrapper } from '@ngxs/store/internals';
+
+export class HmrStoreContext<T, S> {
+  public store: Store;
+  public bootstrap: NgxsBootstrapper;
```
[the remark this answers is not in the document]
Comment (splincode) [reply packages/hmr-plugin/src/internal/hmr-store-context.ts]: done

**Comment by arturovt** [comment]
@markwhitfeld @eranshmil gentlemen :rocket:

File: packages/hmr-plugin/src/actions/hmr-before-destroy.action.ts [hunk 21]
```diff
@@ -0,0 +1,9 @@
+import { NgxsHmrSnapshot } from '../symbols';
+
+export class HmrBeforeDestroyAction {
+  static get type() {
```
Comment (eranshmil) [inline packages/hmr-plugin/src/actions/hmr-before-destroy.action.ts]: Please add this disclaimer above the return statement:
```ts
// NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
```

File: packages/hmr-plugin/src/actions/hmr-init.action.ts [hunk 22]
```diff
@@ -0,0 +1,9 @@
+import { NgxsHmrSnapshot } from '../symbols';
+
+export class HmrInitAction {
+  static get type() {
+    return '@@HMR_INIT';
```
Comment (eranshmil) [inline packages/hmr-plugin/src/actions/hmr-init.action.ts]: Please add this disclaimer above the return statement:
```ts
// NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
```

File: packages/hmr-plugin/src/hmr-manager.ts [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+    this.lifecycle = this.createLifecycle();
+
+    tick();
+
+    return await hmrModule(this.ngModule, this.module);
+  }
+
+  private createLifecycle(): HmrLifecycle<T, S> {
+    const ngAppModule = this.ngModule.instance;
+    const storage = this.storage;
```
Comment (eranshmil) [inline packages/hmr-plugin/src/hmr-manager.ts]: Use destruct for storage, context and optionsBuilder.

File: packages/hmr-plugin/src/internal/hmr-lifecycle.ts [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+    ctx.dispatch(new HmrBeforeDestroyAction(state));
+    return state;
+  }
+
+  private stateEventLoop(frame: HmrCallback<S>) {
+    if (this.storage.existHmrStorage) {
+      const appBootstrapped$: Observable<unknown> = this.bootstrap.appBootstrapped$;
+      const state$: Observable<unknown> = this.context.store.select(state => state);
+
+      appBootstrapped$.subscribe(() => {
```
Comment (eranshmil) [inline packages/hmr-plugin/src/internal/hmr-lifecycle.ts]: Isn't that a memory leak?

File: packages/hmr-plugin/src/internal/hmr-lifecycle.ts [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+    if (typeof this.ngAppModule.hmrNgxsStoreBeforeOnDestroy === 'function') {
+      state = this.ngAppModule.hmrNgxsStoreBeforeOnDestroy(ctx);
+    }
+
+    ctx.dispatch(new HmrBeforeDestroyAction(state));
+    return state;
+  }
+
+  private stateEventLoop(frame: HmrCallback<S>) {
+    if (this.storage.existHmrStorage) {
```
Comment (eranshmil) [inline packages/hmr-plugin/src/internal/hmr-lifecycle.ts]: Please change to:
```ts
if (!this.storage.existHmrStorage) {
  return;
}

...
```

File: packages/hmr-plugin/src/internal/hmr-state-context-factory.ts [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+    if (!store) {
+      throw new Error('Store not found, maybe you forgot to import the NgxsModule');
+    }
+
+    this.store = store;
+  }
+
+  /**
+   * @description
+   * must be taken out into  @ngxs/store/internals
```
Comment (eranshmil) [inline packages/hmr-plugin/src/internal/hmr-state-context-factory.ts]: Are you leaving this for a future PR?

File: docs/plugins/hmr.md [hunk 27]
```diff
@@ -167,3 +163,27 @@ NOTICE Hot Module Replacement (HMR) is enabled for the dev server.
 ```
 
 Now if you make changes to one of your components, those changes should be visible automatically without a complete browser refresh.
+
+### HMR lifecycle
+
+If you want to do some additional manipulations with the states during the hmr lifecycle, you can use the built-in actions. However, this does not work for production.
```
[the remark this answers is not in the document]
Comment (markwhitfeld) [reply docs/plugins/hmr.md]:
```suggestion
If you want to do some modifications to the state during the hmr lifecycle you can use these built-in actions. They will not be executed in production.
```

File: packages/hmr-plugin/src/hmr-manager.ts (hunk 23 above)
Comment (markwhitfeld) [reply packages/hmr-plugin/src/hmr-manager.ts]: Doesn't matter to me. Do what looks better :-)

File: packages/hmr-plugin/src/internal/hmr-lifecycle.ts (hunk 25 above)
Comment (markwhitfeld) [reply packages/hmr-plugin/src/internal/hmr-lifecycle.ts]: Yes, I think that a guard clause here would be cleaner.

File: packages/hmr-plugin/src/internal/hmr-state-context-factory.ts (hunk 26 above)
Comment (markwhitfeld) [reply packages/hmr-plugin/src/internal/hmr-state-context-factory.ts]: Could be left until some future refactoring.

File: integration/app/counter/counter.actions.ts [hunk 28]
```diff
@@ -0,0 +1,3 @@
+export class CounterStateChangeAction {
+  public static readonly type = '[Counter] Change';
```
[the remark this answers is not in the document]
Comment (splincode) [reply integration/app/counter/counter.actions.ts]: in fact, the documentation assumes that the type should be public, it is just an explicit indication

File: packages/hmr-plugin/src/internal/hmr-state-context-factory.ts (hunk 26 above)
Comment (splincode) [reply packages/hmr-plugin/src/internal/hmr-state-context-factory.ts]: Any PR

File: packages/hmr-plugin/src/internal/hmr-lifecycle.ts (hunk 25 above)
Comment (splincode) [reply packages/hmr-plugin/src/internal/hmr-lifecycle.ts]: Done

File: packages/hmr-plugin/src/hmr-manager.ts (hunk 23 above)
Comment (splincode) [reply packages/hmr-plugin/src/hmr-manager.ts]: don't quite understand what you want

File: packages/hmr-plugin/src/actions/hmr-init.action.ts (hunk 22 above)
Comment (splincode) [reply packages/hmr-plugin/src/actions/hmr-init.action.ts]: done

**Comment by splincode** [comment]
@markwhitfeld @eranshmil done

**Comment by splincode** [comment]
@markwhitfeld @eranshmil  ping

**Comment by markwhitfeld** [comment]
@eranshmil Just left for your approval

**Comment by markwhitfeld** [comment]
Oh... I see it is merged. @eranshmil let us know if you are happy or if there is still anything outstanding?

File: integration/app/counter/counter.actions.ts (hunk 28 above)
[the remark this answers is not in the document]
Comment (eranshmil) [reply integration/app/counter/counter.actions.ts]: But we should be consistent.

File: packages/hmr-plugin/src/actions/hmr-before-destroy.action.ts (hunk 21 above)
Comment (eranshmil) [reply packages/hmr-plugin/src/actions/hmr-before-destroy.action.ts]: I've seen you changed that to a property.
You need to change all the actions to the previous definition, please read the PR:
https://github.com/ngxs/store/pull/644

File: packages/hmr-plugin/src/internal/hmr-lifecycle.ts (hunk 25 above)
Comment (eranshmil) [reply packages/hmr-plugin/src/internal/hmr-lifecycle.ts]: I unresolved this only just you won't miss my comment.
Personally, I prefer always using `{}`, even for one line statement.

File: packages/hmr-plugin/src/hmr-manager.ts (hunk 23 above)
Comment (eranshmil) [reply packages/hmr-plugin/src/hmr-manager.ts]: I unresolved this only just you won't miss my comment.
I meant:
```ts
const { storage, ... } = this;
```

**Comment by eranshmil** [comment]
@markwhitfeld Sorry for the late response.
I've replied to 3 comments, only one of them is important:
https://github.com/ngxs/store/pull/765#discussion_r253623750
